### DISCUSSION TITLE
PorousFlow: Problem in convergence

### DISCUSSION THREAD

**Discussion by sinaomrani96** [idaholab/moose#24235] [Q&A Modules: General]
Dear all,
Hope you are doing well.

I wanted to simulate a 2-phase 3-component (water + tracer + CO2) system. First, I simulated the water + tracer system with PorousFlowFullySaturated similar to an example in tutorial_06. The I used "Restarting and recovering from previous simulations" to inject CO2, as I want to have tracer transport as CO2 get injected in my system. However, I have problem regarding convergence. I don't know that this problem is highly nonlinear or there are simpler ways to do it. I appreciate your insight on this. Also, Here are the input files if you want to have a look. Thanks.

https://github.com/sinaomrani96/Test

Best regards,
Sina

**Comment by WilkAndy** [comment]
Hey @sinaomrani96 ,

There is general advice here: https://mooseframework.inl.gov/modules/porous_flow/nonlinear_convergence_problems.html  Please read carefully, as almost every sentence there comes from people struggling with many convergence problems.

Your model is the most difficult situation (multiphase and multicomponent).  It is very easy for PorousFlow to drive a model to a situation where:

- the boundary conditions are trying to remove fluid components that don't exist at the boundary 
- the saturation or mass fraction is very close to zero or one
- the Jacobian is almost singular

You could try running for a while, recording residuals into AuxVars and visualising them to determine what the problem is.  You could try a different initialisation, eg, instead of mass_fraction = 0 somewhere, make it 0.1, just to see what happens.  You could try using a PP (2 porepressures) rather than PS formulation.   You could try changing "variable = " the Kernels+BCs+Diracs , eg, swap "variable = tracer" for "variable = sgas".   You could try turning off various BCs, or adding others (such as removing other components from boundary=right) or using different `use_*` to handle strange cases of saturation=0, etc.

**Reply by sinaomrani96** [reply]
Dear @WilkAndy,
Thanks for your reply.

I will follow your comment and instructions. Sorry to take more of your time but I tried to run the first simulation a bit shorter, and after that run the second simulation (2-phase 3-components) and it converges at least but to a wrong answer (the previous setup failed to converge from the beginning). I tried to record residuals into AuxVars as you suggested but I face Segmentation fault. I was wondering if I am doing it wrong or there is something else. I would appreciate it if you could help me why I cannot record the residuals. Also here is the saturation profile if I turn off the DebugResidualAux.

input files: https://github.com/sinaomrani96/Test/tree/main
![saturatin](https://user-images.githubusercontent.com/64276459/235643367-60e52a92-a99e-45da-9bdc-3a445b116a75.png)

**Reply by WilkAndy** [reply]
Use the `save_in` feature of a `Kernel` to shove the residual into an AuxVariable.

Oh dear, the saturation is waaay wrong (negative saturation).   I suspect it's due to using `PorousFlowCapillaryPressureConst` instead of an experimentally-measured capillary curve.

One other thing to try - probably best to do this before the other items mentioned above - is to copy almost exactly porous_flow/examples/restart/gas_injection.i .  See https://mooseframework.inl.gov/modules/porous_flow/restart.html .    That does almost exactly what you want, but is only 2-component.  Then you can figure out whether it's the extra fluid component or the gas phase that is causing your problems.

**Reply by sinaomrani96** [reply]
Dear @WilkAndy,

Thanks for your help and time. I finally got it running to a more logical answer, but still it needs correction. As I was getting values other than 0 for mass fraction of my second component (which is gas in the gas phase and should not have any value in the 1st phase), I suspected that this is the issue. I do not know if this is a silly approach or actually smart but this is what I did:

As I initialise my second simulation based on a single-phase tracer+water system, I assigned a ParseMaterial to have the value of mass fraction for water in that system. Then I used this for my second run instead of initialising water mass fraction with the values calculated by PorousFlowPropertyAux. However, the problem of getting values (highly negative) for my second component in first phase still persists, and this makes error in calculating mass fractions. How I can prevent this? Here are my input files and results (gas saturation profile and its mass fraction in first phase) now:

https://github.com/sinaomrani96/Test/tree/main


![saturatin](https://user-images.githubusercontent.com/64276459/235910531-fd72fa08-89b8-44c7-9446-43c379da6442.png)
![x02](https://user-images.githubusercontent.com/64276459/235910655-ff7f77f6-459a-4e5c-a453-15287d25ff4b.png)


Thanks.

**Reply by WilkAndy** [reply]
Hey @sinaomrani96 - i'm not going to have time to look at this for the next few days.   i'm giving some important lectures and organising a conference and they're taking all my time.

**Reply by sinaomrani96** [reply]
Hi,

I just wanted to update the status of this problem that changing the order of numbering will solve this issue. Thanks.
